library/waku_thread/waku_thread.nim

```diff
@@ -0,0 +1,188 @@
+    locals = addr(locals)
+    nimGC_setStackBottom(locals)
+
+proc createNode(configJson: cstring): Result[WakuNode, string] =
+  var privateKey: PrivateKey
+  var netConfig = NetConfig.init(ValidIpAddress.init("127.0.0.1"),
+                                 Port(60000'u16)).value
+  var relay: bool
+  var topics = @[""]
+  var jsonResp: JsonEvent
+
+  let cj = configJson.alloc()
+
+  if not parseConfig($cj,
+                     privateKey,
+                     netConfig,
+                     relay,
+                     topics,
+                     jsonResp):
+    deallocShared(cj)
+    return err($jsonResp)
+
+  deallocShared(cj)
+
+  var enrBuilder = EnrBuilder.init(privateKey)
+
+  enrBuilder.withIpAddressAndPorts(
+    netConfig.enrIp,
+    netConfig.enrPort,
+    netConfig.discv5UdpPort
+  )
+
+  if netConfig.wakuFlags.isSome():
+    enrBuilder.withWakuCapabilities(netConfig.wakuFlags.get())
+
+  enrBuilder.withMultiaddrs(netConfig.enrMultiaddrs)
+
+  let addShardedTopics = enrBuilder.withShardedTopics(topics)
+  if addShardedTopics.isErr():
+    let msg = "Error setting shared topics: " & $addShardedTopics.error
+    return err($JsonErrorEvent.new(msg))
+
+  let recordRes = enrBuilder.build()
+  let record =
+    if recordRes.isErr():
+      let msg = "Error building enr record: " & $recordRes.error
+      return err($JsonErrorEvent.new(msg))
+
+    else: recordRes.get()
+
+  var builder = WakuNodeBuilder.init()
+  builder.withRng(crypto.newRng())
+  builder.withNodeKey(privateKey)
+  builder.withRecord(record)
+  builder.withNetworkConfiguration(netConfig)
+  builder.withSwitchConfiguration(
+    maxConnections = some(50.int)
+  )
+
+  let wakuNodeRes = builder.build()
+  if wakuNodeRes.isErr():
+    let errorMsg = "failed to create waku node instance: " & wakuNodeRes.error
+    return err($JsonErrorEvent.new(errorMsg))
+
+  var newNode = wakuNodeRes.get()
+
+  if relay:
+    waitFor newNode.mountRelay()
+    newNode.peerManager.start()
+
+  return ok(newNode)
+
+proc run(ctx: ptr Context) {.thread.} =
+  ## This is the worker thread body. This thread runs the Waku node
+  ## and attends library user requests (stop, connect_to, etc.)
+
+  while running.load == true:
+    ## Trying to get a request from the libwaku main thread
+    let req = ctx.reqChannel.tryRecv()
+    if req[0] == true:
+      let response = waitFor req[1].process(ctx.node)
+      ctx.respChannel.send( response )
+
+    poll()
+
+  tearDownForeignThreadGc()
+
+proc createWakuThread*(configJson: cstring): Result[void, string] =
+  ## This proc is called from the main thread and it creates
+  ## the Waku working thread.
+
+  waku_init()
+
+  ctx = createShared(Context, 1)
+  ctx.reqChannel.open()
+  ctx.respChannel.open()
+
+  let newNodeRes = createNode(configJson)
+  if newNodeRes.isErr():
+    return err(newNodeRes.error)
+
+  ctx.node = newNodeRes.get()
+
+  running.store(true)
+
+  try:
+    createThread(ctx.thread, run, ctx)
```

Comment:
the context contains garbage-collected types - therefore, it is not safe to pass it to a different thread where references to these instances may accidentally be copied.

In particualr, the `node` itself must not be allocated in this thread and then used in the other thread - this will lead to memory corruption.

Instead, the Node instance should be created inside `run`, in the thread where it's running - to configure it, a configuration should be passed to it that is allocated with allocShared and doesn't contain any refcounted types.

The rule is: never allocate anything with `new` (no ref object, etc) in one thread and use it in the other.

Reply:
To elaborate, `Context` should only contain thread-safe types, ie those that are _not_ `string`, `seq,` and `ref`. It can contain `ptr`, `array` and other "manually" allocated types.